2022/01/how-to-use-regex-groups-quantifier-in-postgresql.md

```diff
@@ -0,0 +1,86 @@
+---
+author: "Selvakumar Arumugm"
+title: "How to use regex groups quantifier in PostgreSQL"
+date: 2022-01-07
+tags:
+- postgres
+- sql
+- regex
+- hl7
+github_issue_number: 
+---
+
+I recently encountered a situation where it was necessary to extract address content from a
+specific format (HL7 V2) of text on a PostgreSQL table column. This became possible through applying a regex as the address contains a standard format that regex can match. The following example will be used to demonstrate the process: 
+
+```
+||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|
+```
+### Perl Regex Pattern
+
+In order to manipulate our example, the address section needs to be extracted from the table. The sections and fields are delimited by pipes ( | ) and carets ( ^ ). Uniquely, the address contains eight carets. Therefore, a regex group with alphanumeric and caret repetition is required in order to match our address. The regex is tested through the grep command perl regex option (-P, --perl-regexp) and successfully extracts the address section from the content.
```

Comment:
@jonjensen I didn't cover much on HL7 message since I was focusing on Postgres feature. But I have planned to write a HL7 V2 specific blog post explaining the concepts of HL7 V2 message. I will refer that blog here once that ready. I am hoping to get that one ready by this month.

Reply:
Ok. That makes sense, and I don't want this to be a big explanation of HL7.

But I feel like we need to say enough about the HL7 data here to understand what the regex is doing.

Can you at least explain the difference between `|` and `^` in HL7? And whether your pattern of `[A-Za-z0-9 ]` is really right for matching, since it wouldn't allow names like O'Leary or Jean-Marie?

Those two little things are probably enough for readers to make sense of this.

For me, otherwise, I wonder if the regex will exclude legitimate matches, and the regex *is* your point here.

Reply:
@jonjensen sorry for delay. I have added additional notes on HL7 V2 message delimiters and added more regex coverage to address special characters.